### ISSUE TITLE
Rule Change: There should be a way to specify ignores for no-param-reassign

### ISSUE THREAD

**Issue by DamienCassou** [eslint/eslint#16367]
### What rule do you want to change?

no-param-reassign

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
function myWidget(spec) {
	spec ||= {};
	...
}
```

### What does the rule currently do for this code?

Report that `spec` is reassigned.

### What will the rule do after it's changed?

With `{ignore: ["spec"]}`, the rule should stop reporting reassignments to `spec`.

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @DamienCassou, thanks for the issue!

Can you provide more details on the motivation for this option? In particular, why should parameter reassignment only be allowed for certain names in an entire codebase?

**Comment by DamienCassou** [comment]
Hi @mdjermanovic. Thank you for reading my issue and asking questions.

> Can you provide more details on the motivation for this option?

We have a lot of code looking like:

```js
function myClass(spec, my) {
	spec ||= {};
	my ||= {};
	...
}
```

This kind of (old-fashioned) code comes from the book "JavaScript: The Good Parts" by [Douglas Crockford](https://en.wikipedia.org/wiki/Douglas_Crockford) (author of JSON).

Our code base has 423 problems detected by `no-param-reassign` including 168 with the variable `my` and 183 with spec. Adding an `ignore` pattern to `no-param-reassign` will allow us to focus on fixing 72 real problems (72=423-168-183).

Does it make sense?

**Comment by snitin315** [comment]
> Our code base has 423 problems detected by no-param-reassign including 168 with the variable my and 183 with spec. Adding an ignore pattern to no-param-reassign will allow us to focus on fixing 72 real problems (72=423-168-183).

I feel it makes sense to allow users to ignore a few variables if they want to 👍🏻

**Comment by DamienCassou** [comment]
Please tell me when I can start working on a PR.

**Comment by mdjermanovic** [comment]
Sorry for the delay.

I'm not in favor of adding this option to the core rule because I think there should be no reason to allow exceptions by param names in a new codebase. In particular, it would be confusing why the option exists and could lead to logic errors in the code, as in https://github.com/eslint/eslint/issues/15244. For specific use cases, such as an existing codebase that follows a certain pattern, it's always possible to create a custom rule based on the core rule.

**Comment by DamienCassou** [comment]
@mdjermanovic I see your point.

> it's always possible to create a custom rule based on the core rule.

Do you have an example as to how doing that beyond copy/paste/edit?

Feel free to close the issue.

**Comment by mdjermanovic** [comment]
There are tools to wrap and extend existing rules, like [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) and [eslint-rule-extender](https://github.com/kaicataldo/eslint-rule-extender), but note that rule definitions are not public API and therefore can change and possibly break extension rules in non-major versions. Copy & paste & edit is actually the recommended way.
